File: rows_test.go
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package spannerdriver

import (
	"database/sql/driver"
	"fmt"
	"io"
	"testing"

	"cloud.google.com/go/spanner"
	sppb "cloud.google.com/go/spanner/apiv1/spannerpb"
	"google.golang.org/protobuf/types/known/structpb"
)

type testIterator struct {
	metadata *sppb.ResultSetMetadata
	rows     []*spanner.Row
	index    int
}

func (t *testIterator) Next() (*spanner.Row, error) {
	if t.index == len(t.rows) {
		return nil, io.EOF
	}
	row := t.rows[t.index]
	t.index++
	return row, nil
}

func (t *testIterator) Stop() {
}

func (t *testIterator) Metadata() *sppb.ResultSetMetadata {
	return t.metadata
}

func newRow(t *testing.T, cols []string, vals []interface{}) *spanner.Row {
	row, err := spanner.NewRow(cols, vals)
	if err != nil {
		t.Fatalf("failed to create test row: %v", err)
	}
	return row
}

func TestRows_Next(t *testing.T) {
	cols := []string{"COL1", "COL2", "COL3"}
	it := testIterator{
		metadata: &sppb.ResultSetMetadata{
			RowType: &sppb.StructType{
				Fields: []*sppb.StructType_Field{
					{Name: "COL1", Type: &sppb.Type{Code: sppb.TypeCode_INT64}},
					{Name: "COL2", Type: &sppb.Type{Code: sppb.TypeCode_STRING}},
					{Name: "COL3", Type: &sppb.Type{Code: sppb.TypeCode_FLOAT64}},
				},
			},
		},
		rows: []*spanner.Row{
			newRow(t, cols, []interface{}{int64(1), "Test1", 3.14 * 1.0}),
			newRow(t, cols, []interface{}{int64(2), spanner.NullString{}, spanner.NullFloat64{}}),
			newRow(t, cols, []interface{}{int64(3), "Test3", 3.14 * 3.0}),
		},
	}

	rows := rows{it: &it}
	dest := make([]driver.Value, 3)
	values := make([][]driver.Value, 0)
	for {
		err := rows.Next(dest)
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatal(err)
		}
		row := make([]driver.Value, 3)
		copy(row, dest)
		values = append(values, row)
	}
	if g, w := len(values), 3; g != w {
		t.Fatalf("values length mismatch\nGot: %v\nWant: %v", g, w)
	}
	for i := int64(0); i < int64(len(values)); i++ {
		if g, w := values[i][0], i+1; g != w {
			t.Fatalf("COL1 value mismatch for row %d\nGot: %v\nWant: %v", i, g, w)
		}
		if i == 0 || i == 2 {
			if g, w := values[i][1], fmt.Sprintf("Test%d", i+1); g != w {
				t.Fatalf("COL2 value mismatch for row %d\nGot: %v\nWant: %v", i, g, w)
			}
			if g, w := values[i][2], 3.14*float64(i+1); g != w {
				t.Fatalf("COL3 value mismatch for row %d\nGot: %v\nWant: %v", i, g, w)
			}
		} else {
			if g, w := values[i][1], driver.Value(nil); g != w {
				t.Fatalf("COL2 value mismatch for row %d\nGot: %v\nWant: %v", i, g, w)
			}
			if g, w := values[i][2], driver.Value(nil); g != w {
				t.Fatalf("COL3 value mismatch for row %d\nGot: %v\nWant: %v", i, g, w)
			}
		}
	}
}

func TestRows_Next_Unsupported(t *testing.T) {
	unspecifiedType := &sppb.Type{Code: sppb.TypeCode_TYPE_CODE_UNSPECIFIED}

	it := testIterator{
		metadata: &sppb.ResultSetMetadata{
			RowType: &sppb.StructType{
				Fields: []*sppb.StructType_Field{
					{Name: "COL1", Type: unspecifiedType},
				},
			},
		},
		rows: []*spanner.Row{
			newRow(t, []string{"COL1"}, []any{
				spanner.GenericColumnValue{
					Type:  unspecifiedType,
					Value: &structpb.Value{},
				},
			}),
		},
	}

	rows := rows{it: &it}

	dest := make([]driver.Value, 1)
	err := rows.Next(dest)
	if err == nil {
		t.Fatal("expected an error, but got nil")
	}
	const expectedError = "unsupported type TYPE_CODE_UNSPECIFIED, use spannerdriver.ExecOptions{DecodeOption: spannerdriver.DecodeOptionProto} to return the underlying protobuf value"
	if err.Error() != expectedError {
		t.Fatalf("expected error %q, but got %q", expectedError, err.Error())
	}
}

File: spannerlib/wrappers/spannerlib-node/test/rows.test.ts
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as assert from 'assert';
import { Rows } from '../src/lib/rows.js';
import { Connection } from '../src/lib/connection.js';
import { Pool } from '../src/lib/pool.js';
import { ffi } from '../src/ffi/utils.js';
import sinon from 'sinon';
import pkg from '@google-cloud/spanner/build/protos/protos.js';
const { google } = pkg;
const ListValue = google.protobuf.ListValue;
const Value = google.protobuf.Value;

describe('Rows', () => {
  let stub: sinon.SinonStub;

  beforeEach(() => {
    stub = sinon.stub(ffi, 'invokeAsync');
  });

  afterEach(() => {
    stub.restore();
  });

  interface MockListValue {
    values: Array<{ stringValue: string }>;
  }

  it('should fetch next row successfully', async () => {
    const pool = new Pool(
      'node-esm',
      'projects/test/instances/test/databases/test'
    );
    pool.oid = 1;
    const connection = new Connection();
    connection.pool = pool;
    connection.oid = 2;

    const rows = new Rows(connection, 3);

    // Create a dummy ListValue
    const listValue = ListValue.create({
      values: [Value.create({ stringValue: '1' })],
    });
    const buffer = ListValue.encode(listValue).finish() as Buffer;

    stub
      .onFirstCall()
      .resolves({ objectId: 0, pinnerId: 0, protobufBytes: buffer });

    const row = (await rows.next()) as unknown as MockListValue;

    assert.ok(row, 'Row should be returned');
    assert.strictEqual(row.values.length, 1, 'Row should have 1 value');
    assert.strictEqual(row.values[0].stringValue, '1', 'Value should be "1"');

    assert.strictEqual(
      stub.calledOnce,
      true,
      'invokeAsync should be called once'
    );
    assert.strictEqual(
      stub.firstCall.args[0],
      'Next',
      'First call should be Next'
    );
  });

  it('should return null when no more rows', async () => {
    const pool = new Pool(
      'node-esm',
      'projects/test/instances/test/databases/test'
    );
    pool.oid = 1;
    const connection = new Connection();
    connection.pool = pool;
    connection.oid = 2;

    const rows = new Rows(connection, 3);

    stub
      .onFirstCall()
      .resolves({ objectId: 0, pinnerId: 0, protobufBytes: null });

    const row = await rows.next();

    assert.strictEqual(row, null, 'Row should be null');
  });

  it('should retrieve metadata successfully', async () => {
    const pool = new Pool(
      'node-esm',
      'projects/test/instances/test/databases/test'
    );
    pool.oid = 1;
    const connection = new Connection();
    connection.pool = pool;
    connection.oid = 2;

    const rows = new Rows(connection, 3);

    const metadataProto = google.spanner.v1.ResultSetMetadata.create({
      rowType: {
        fields: [{ name: 'col1', type: { code: 'INT64' } }],
      },
    });
    const buffer = google.spanner.v1.ResultSetMetadata.encode(
      metadataProto
    ).finish() as Buffer;

    stub
      .onFirstCall()
      .resolves({ objectId: 0, pinnerId: 0, protobufBytes: buffer });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const metadata: any = await rows.metadata();
    assert.ok(metadata, 'Metadata should be returned');
    assert.ok(metadata.rowType, 'RowType should exist');
    assert.strictEqual(metadata.rowType.fields[0].name, 'col1');
    assert.strictEqual(stub.firstCall.args[0], 'Metadata');
  });

  it('should retrieve stats successfully', async () => {
    const pool = new Pool(
      'node-esm',
      'projects/test/instances/test/databases/test'
    );
    pool.oid = 1;
    const connection = new Connection();
    connection.pool = pool;
    connection.oid = 2;

    const rows = new Rows(connection, 3);

    const statsProto = google.spanner.v1.ResultSetStats.create({
      rowCountExact: 5,
    });
    const buffer = google.spanner.v1.ResultSetStats.encode(
      statsProto
    ).finish() as Buffer;

    stub
      .onFirstCall()
      .resolves({ objectId: 0, pinnerId: 0, protobufBytes: buffer });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const stats: any = await rows.resultSetStats();
    assert.ok(stats, 'Stats should be returned');
    assert.strictEqual(Number(stats.rowCountExact), 5);
    assert.strictEqual(stub.firstCall.args[0], 'ResultSetStats');
  });

  it('should advance to next result set successfully', async () => {
    const pool = new Pool(
      'node-esm',
      'projects/test/instances/test/databases/test'
    );
    pool.oid = 1;
    const connection = new Connection();
    connection.pool = pool;
    connection.oid = 2;

    const rows = new Rows(connection, 3);

    const metadataProto = google.spanner.v1.ResultSetMetadata.create({
      rowType: {
        fields: [{ name: 'col2', type: { code: 'STRING' } }],
      },
    });
    const buffer = google.spanner.v1.ResultSetMetadata.encode(
      metadataProto
    ).finish() as Buffer;

    stub
      .onFirstCall()
      .resolves({ objectId: 0, pinnerId: 0, protobufBytes: buffer });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const metadata: any = await rows.nextResultSet();
    assert.ok(metadata, 'Metadata should be returned');
    assert.strictEqual(metadata.rowType.fields[0].name, 'col2');
    assert.strictEqual(stub.firstCall.args[0], 'NextResultSet');
  });

  it('should calculate updateCount correctly for exact and lower bound values', async () => {
    const pool = new Pool(
      'node-esm',
      'projects/test/instances/test/databases/test'
    );
    pool.oid = 1;
    const connection = new Connection();
    connection.pool = pool;
    connection.oid = 2;

    const rows = new Rows(connection, 3);

    // Exact count test
    const exactStats = google.spanner.v1.ResultSetStats.create({
      rowCountExact: 12,
    });
    const exactBuffer = google.spanner.v1.ResultSetStats.encode(
      exactStats
    ).finish() as Buffer;

    stub
      .onFirstCall()
      .resolves({ objectId: 0, pinnerId: 0, protobufBytes: exactBuffer });
    const countExact = await rows.updateCount();
    assert.strictEqual(countExact, 12);

    stub.restore();
    stub = sinon.stub(ffi, 'invokeAsync');

    // Lower bound count test
    const lowerBoundStats = google.spanner.v1.ResultSetStats.create({
      rowCountLowerBound: 42,
    });
    const lowerBoundBuffer = google.spanner.v1.ResultSetStats.encode(
      lowerBoundStats
    ).finish() as Buffer;

    stub
      .onFirstCall()
      .resolves({ objectId: 0, pinnerId: 0, protobufBytes: lowerBoundBuffer });
    const countLower = await rows.updateCount();
    assert.strictEqual(countLower, 42);
  });

  it('should throw error on operations if connection is closed', async () => {
    const pool = new Pool(
      'node-esm',
      'projects/test/instances/test/databases/test'
    );
    pool.oid = 1;
    const connection = new Connection();
    connection.pool = pool;
    connection.oid = 2;

    const rows = new Rows(connection, 3);
    connection.closed = true;

    await assert.rejects(async () => {
      await rows.next();
    }, /Connection is closed or invalid/);

    await assert.rejects(async () => {
      await rows.metadata();
    }, /Connection is closed or invalid/);

    await assert.rejects(async () => {
      await rows.resultSetStats();
    }, /Connection is closed or invalid/);

    await assert.rejects(async () => {
      await rows.nextResultSet();
    }, /Connection is closed or invalid/);
  });

  it('should throw error on operations if rows are closed', async () => {
    const pool = new Pool(
      'node-esm',
      'projects/test/instances/test/databases/test'
    );
    pool.oid = 1;
    const connection = new Connection();
    connection.pool = pool;
    connection.oid = 2;

    const rows = new Rows(connection, 3);
    rows.closed = true;

    await assert.rejects(async () => {
      await rows.next();
    }, /Rows are already closed/);

    await assert.rejects(async () => {
      await rows.metadata();
    }, /Rows are already closed/);

    await assert.rejects(async () => {
      await rows.resultSetStats();
    }, /Rows are already closed/);

    await assert.rejects(async () => {
      await rows.nextResultSet();
    }, /Rows are already closed/);
  });
});
